content/post/hercules.v4.md

```diff
@@ -0,0 +1,295 @@
+
+If you haven't read [Your Code as a Crime Scene](https://pragprog.com/book/atcrime/your-code-as-a-crime-scene)
+by Adam Tornhill, stop reading this post, allocate a few hours and look it through - the book is awesome.
+One of the ideas from the book is to investigate which files are _coupled_ - often appear together.
+The corresponding analysis in Hercules goes a bit further and puts all the files
+into the 3D space with the property that the distance between coupled files is small.
+
+{{% caption src="/post/hercules.v4/couples.png" %}}
+Files coupling in [tensorflow/tensorflow](https://github.com/tensorflow/tensorflow). [Click here for an interactive experience.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_files.json)
+{{% /caption %}}
+
+Reproduce:
+```
+hercules --couples --pb https://github.com/tensorflow/tensorflow | labours.py -m couples -o tensorflow
+```
+
+We can repeat the same trick with the developers: they are couples if they commit
+to the same files. [Interactive example on Tensorflow developers.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_people.json)
+
+Finally, the trick works once again with structural units in the source code. We parse
+each file with [Babelfish](https://doc.bblf.sh) and extract anything we wish: functions,
+classes, etc. Then we apply the same coupling analysis to them: we know which are changed
+in each commit.
+
+By default, the functions are extracted.
+```
+hercules --shotness --pb https://github.com/tensorflow/tensorflow | labours.py -m shotness -o tensorflow
+```
+
+#### Comment sentiment
+
+This is a proof-of-concept for running a Tensorflow model over the source code.
+We take [BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) and apply it to comments.
+That model is general-purpose and the result is often weird, but it works.
+I wrote about it in [the other blog post](../codesent).
+
+Your binary must be compiled with Tensorflow support (the released one is **not**).
+```
+hercules --sentiment --pb https://github.com/tensorflow/tensorflow | labours.py -m sentiment
+```
+
+# v4 architecture
+
+Hercules is built on top of the `Pipeline` which is a Directed Acyclic Graph (DAG) of `PipelineItem`-s.
+Each pipeline item has a name and specifies its dependencies and what it provides in return.
+For example, a `RenameAnalysis` item detects the renamed and slightly changed files between two commits.
+It consumes the list of changes and the list of corresponding blobs. It calculates
+a better list of changes with some of the "delete + create" pairs replaced with
+single "edit under new name" elements. Yep, Git packfiles do not store this kind
+of information (intentionally) and `git` detects renames using heuristics every
+time you execute it.
+
+Hercules builds the DAG of pipeline items automatically based on their dependencies
+and provided intermediate results. Some items try to improve on the others -
```

Comment:
This sentence needs some work, it's too complex and hard to parse.
Also, what do you mean the items try to improve on the others? Which others?

Maybe:

> Each item on a pipeline transforms the result of its dependencies to generate its own result,
> as `RenameAnalysis` mentioned above.
> Hercules understands these dependencies and automatically sets up the corresponding computation graph.
> Here is an example of the built pipeline to calculate the burn down chart.

Reply:
This simplifies the things but hell yeah better to simplify than to make more complex